src/Npgsql/Replication/NpgsqlReplicationConnection.cs

```diff
@@ -0,0 +1,888 @@
+                    : ParseByteaEscape(bytes);
+
+                byte[] ParseByteaHex(ReadOnlySpan<byte> inBytes)
+                {
+                    var outBytes = new byte[inBytes.Length / 2];
+                    for (var i = 0; i < inBytes.Length; i++)
+                    {
+                        var v1 = inBytes[i++];
+                        var v2 = inBytes[i];
+                        outBytes[i / 2] =
+                            (byte)(((v1 - (v1 < 0x3A ? 0x30 : 87)) << 4) | (v2 - (v2 < 0x3A ? 0x30 : 87)));
+                    }
+
+                    return outBytes;
+                }
+
+                byte[] ParseByteaEscape(ReadOnlySpan<byte> inBytes)
+                {
+                    var result = new MemoryStream(new byte[inBytes.Length]);
+                    for (var tp = 0; tp < inBytes.Length;)
+                    {
+                        var c1 = inBytes[tp];
+                        if (c1 != '\\')
+                        {
+                            // Don't validate whether c1 >= 0x20 && c1 <= 0x7e here
+                            // TIMELINE_HISTORY currently (2020-09-13) sends raw
+                            // bytes instead of bytea for the content value.
+                            result.WriteByte(c1);
+                            tp++;
+                            continue;
+                        }
+
+                        var c2 = inBytes[tp + 1];
+                        if (c2 == '\\')
+                        {
+                            result.WriteByte(c2);
+                            tp += 2;
+                            continue;
+                        }
+
+                        var c3 = inBytes[tp + 2];
+                        var c4 = inBytes[tp + 3];
+                        if (c2 >= '0' && c2 <= '3' &&
+                            c3 >= '0' && c3 <= '7' &&
+                            c4 >= '0' && c4 <= '7')
+                        {
+                            c2 <<= 3;
+                            c2 += c3;
+                            c2 <<= 3;
+                            result.WriteByte((byte)(c2 + c4));
+
+                            tp += 4;
+                            continue;
+                        }
+
+                        throw new FormatException("Invalid syntax for type bytea");
+                    }
+
+                    return result.ToArray();
+                }
+            }
+        }
+
+        void SetTimeouts(TimeSpan readTimeout, TimeSpan writeTimeout)
+        {
+            var connector = Connector;
+            connector.UserTimeout = readTimeout > TimeSpan.Zero ? (int)readTimeout.TotalMilliseconds : 0;
+            connector.WriteBuffer.Timeout = writeTimeout;
+        }
+
+        void Cancel()
+        {
+            _sendFeedbackTimer.Change(Timeout.InfiniteTimeSpan, Timeout.InfiniteTimeSpan);
+            _requestFeedbackTimer.Change(Timeout.InfiniteTimeSpan, Timeout.InfiniteTimeSpan);
+            Connector.CancelRequest();
```

Comment:
Isn't it possible here to disable the timers, but the timer's callbacks will turn them on? (Same thing for `SendStatusUpdate`)

Reply:
Yep, that seems like a race condition.

@Brar, since cancellation is out-of-band with another connection, why do we need to disable the timers btw?

Reply:
> Isn't it possible here to disable the timers, but the timer's callbacks will turn them on? 

Yes, we need to make sure that the timers aren't re-enabled by the callback.
This was the thing the Cancelling state that @roji removed was meant to prevent. There was still a race condition left though. 

> Same thing for `SendStatusUpdate`

I don't think this could happen as both `SendStatusUpdate` and `TimerSendFeedback` wait on the _feedbackSemaphore - no?

> @Brar, since cancellation is out-of-band with another connection, why do we need to disable the timers btw?

Because the backend is only prepared to receive feedback while streaming. If the timers would keep on sending feedback after successful cancellation we'd probably loose protocol sync.

Reply:
> I don't think this could happen as both SendStatusUpdate and TimerSendFeedback wait on the _feedbackSemaphore - no?

I was talking about calling `SendStatusUpdate` at the same time as `Cancel`.

Reply:
Ok, yes. Cancel races and is somewhat problematic because there always is the question if/how long we're willing to block in there to avoid race conditions.
For instance we may get stuck in the callback when there's a network partition just before sending feedback.

Reply:
https://github.com/Brar/npgsql/pull/10